src/operator/contrib/krprod.h

```diff
@@ -0,0 +1,202 @@
+  }
+  CHECK_EQ(kronecker_length, static_cast<int>(out.size(1)));
+
+  // Create an intermediate space of the same shape as out
+  //
+  // Suppose storage stores the result at step i-1, we'd
+  // compute and store the result into out for step i;
+  // we then proceed to compute and store the result in storage
+  // for step i+1 and so on and so forth, by alternating using
+  // storage and out to store the given variable and the result variable
+  Tensor<cpu, 2, DType> storage(out.shape_);
+  AllocSpace(&storage);
+
+  // Pointers to the given variable and result variable
+  // We exchange what given and result point to at every step
+  Tensor<cpu, 2, DType> *given = &storage,
+    *result = &out, *tmp;
+
+  // Compute each intermediate Khatri-Rao product
+  storage = 1;
+  kronecker_length = 1;
+  for (auto & ts : ts_arr) {
+    expr::BLASEngine<cpu, DType>::SetStream
+      (result->stream_);
+
+    // Compute the current Khatri-Rao product, row by row
+    *result = 0;
+    for (int i = 0; i < nrows; ++i) {
+      // BLAS signature
+      //
+      // dger(
+      //   m : ts.size(1), length of each row of current matrix
+      //   n : kronecker_length, length of each row of previous result
+      //   alpha : 1, scaling to the outer product of x and y
+      //   x : ts[i].dptr_, current row of current matrix
+      //   incx : 1, as each element in the row is contiguous
+      //   y : (*given)[i].dptr_, current row of the given variable
+      //   incy : 1, as each element in the row is contiguous
+      //   a : (*result)[i].dptr_, current row of the result variable
+      //   lda : ts.size(1), as the outer product is stored as one row
+      //         which occupies contiguous memory, and as BLASEngine::ger()
+      //         assumes column-major matrix, lda has to be precisely
+      //         the length of x, i.e. ts[i].size(1)
+      // )
+      expr::BLASEngine<cpu, DType>::ger
+        (result->stream_,
+        ts.size(1), kronecker_length, 1,
+        ts[i].dptr_, 1,
+        (*given)[i].dptr_, 1,
+        (*result)[i].dptr_, ts.size(1));
+    }
+    kronecker_length *= ts.size(1);
+
+    tmp = given;
+    given = result;
+    result = tmp;
+  }
+
+  // If the final result is stored in storage,
+  // copy its value to out
+  if (given != &out)
+    Copy(out, storage);
+
+  FreeSpace(&storage);
+}
+
+/*!
+ * \brief Convenience function for row-wise Kronecker product of two matrices
+ *
+ * \param out result matrix
+ * \param in1 first input matrix
+ * \param in2 second input matrix
+ */
+template <typename DType>
+inline void row_wise_kronecker
+  (Tensor<cpu, 2, DType> out,
+  const Tensor<cpu, 2, DType> &in1,
+  const Tensor<cpu, 2, DType> &in2) {
+  row_wise_kronecker(out, std::vector<Tensor<cpu, 2, DType> > {in1, in2});
+}
+
+/*!
+ * \brief Khatri-Rao product
+ *
+ * \param out result matrix
+ * \param ts_arr vector of input matrices
+ */
+template <typename DType>
+inline void krprod
```

Comment:
The problem is that now we're copying everything first (when transposing). That might be slower than operating directly on the original matrices. Ideally we'd want to have a flag specifying whether the matrices need transposing or not (`gemm` style)